### PR TITLE
Component time tracking

### PR DESCRIPTION
**Pull request by magnusuMET**
Adds timer functionality which is used to track the time spent on individual components of the SNAP model. The python script can be used to plot the time spent both per iteration and per (iteration, particle) during a run without requiring the model to finish execution

### CODE REVIEW

**Comment by heikoklein** [comment]
The functionality is very much needed and it is very nice to have a plotting function in combination with that.

I think the timer just gives currently cpu_times. real-times are at least as import to assess if e.g. parallelization works or if IO-waits are the bottleneck: `timer: Particle loop: YYY real, XXX cpu`

The time-logtext should already be set during initialization, eventually in two parts, e.g.

```
  global_timer_prefix = "timer: " # could be a default
  global_timer_total_prefix = "timer: Total: " # see later for total output
  pl_loop_timer = timer("Particle loop: ")
  io_timer = timer("output/accumulation: ")
```

Log-functions should be parameterized with the log-unit, eventually using iulog as default. E.g.
```
  timer%log(error_unit)
```

Rather than to reset the timer all the time, timers should be started/stopped, and eventually at stop_log functionality (combined stop and log, since that is convenient).
```
  pl_loop_timer%start()
 ...
  pl_loop_timer%stop_log(error_unit)
```

It should be possible to print the accumulated total log:
```
  pl_loop_timer%log_total(output_unit)
```
giving a line like: `timer: Total: Particle loop: YYY real, XXX cpu`

**Comment by magnusuMET** [comment]
For walltime we have the following:
* `system_clock`: Funky frequency scaling issues
* `date_and_time`: Requires parsing of date-times, could be done when post-processing in python
* `dtime`: GNU extension which provides both system and user time

Do you have any preferences in regards to which routines would be best to use here?

**Comment by heikoklein** [comment]
What about `omp_get_wtime`? As a fallback you could use system_clock together with cpu_time.

**Comment by magnusuMET** [comment]
We have a conditional on `_OPENMP`, this is a documented compile definition according to the openmp standard.

**Comment by heikoklein** [comment]
Yes, you will need that when you use omp_get_wtime.

**Review by heikoklein — COMMENTED** [review]
The output 
```timer: output/accumulation:   0: 0: 0.  21 (sys)   0: 0: 0.  21 (cpu)
 timer: time_loop:   0: 0: 0. 136 (sys)   0: 0: 0. 135 (cpu)
 timer: Total: time_loop:   0: 4:50. 345 (sys)   0: 2:27. 517 (cpu)
 timer: Total: output/accumulation:   0: 0: 5. 198 (sys)   0: 0: 5. 201 (cpu)
 timer: Total: Reading MET input:   0: 2:53. 707 (sys)   0: 0:15. 798 (cpu)
 timer: Total: Particle loop:   0: 2: 4. 320 (sys)   0: 2: 4. 321 (cpu)
```
is difficult to read. Times should be given as 0:02:04.320 (or is it .0320?), i.e. H:MM:SS.MS.
`(I3, a, I0.2, a, I0.2, a, I0.4)`. Can all Total times be written under each others, i.e. by giving the text before the timings a fixed format a35?

sys is usually (time-command) the time used by system/kernel. What you call sys is usually called real or wall time: https://stackoverflow.com/questions/556405/what-do-real-user-and-sys-mean-in-the-output-of-time1

File: src/common/snapdebugML.F90 [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+#endif
+
+    function to_str(t) result(str)
+      real, intent(in) :: t
+      character(len=64) :: str
+
+      integer :: h, m, s, ms
+      call real_to_hms_ms(t, h, m, s, ms)
+
+      write(str,"(I3, a, I2, a, I2, a, I4)") h, ":", m, ":", s, ".", ms
```
Comment (heikoklein) [inline src/common/snapdebugML.F90:175]: might be more elegant as
`write(str,"(I3, ':', I0.2, ':', I0.2, '.', I04)") h, m, s, ms`

Comment (heikoklein) [reply src/common/snapdebugML.F90:175]: `I0.4`

**Comment by magnusuMET** [comment]
An early return gave non-sensible `time_loop` times, this is fixed in the latest commit. We should see if we can interleave the reading of the next MET field with particle computations. The differences between `wall` and `cpu` is quite telling for how many cycles we are losing

**Comment by heikoklein** [comment]
Looks nearly perfect, but the timer-line is now 81 characters long, one character too much for my terminal. And two whitespace in the beginning. Would be good to restrict it to <=80chars somehow
```
  timer: output/accumulation:              0:00:00.021 (wall)   0:00:00.021 (cpu
)
  timer: time_loop:                        0:00:00.136 (wall)   0:00:00.136 (cpu
)
  timer: Total: time_loop:                 0:05:15.145 (wall)   0:02:35.876 (cpu
)
  timer: Total: output/accumulation:       0:00:05.504 (wall)   0:00:05.501 (cpu
)
  timer: Total: Reading MET input:         0:02:54.933 (wall)   0:00:15.675 (cpu
)
  timer: Total: Particle loop:             0:02:10.683 (wall)   0:02:10.665 (cpu
)
```

**Review by heikoklein — APPROVED** [review]
Thanks, please merge. Very good information.

**Pull request merged**